doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,230 @@
+
+## Goals
+
+The goal of this design document is to describe what changes are needed to provide Strict Serializability to the
+operations mentioned in correctness property 1.
+
+## Non-Goals
+
+Sinks are left out of this document and not considered for consistency.
+
+`TAIL` is left out of this document and not considered for consistency.
+
+`AS OF` allows clients to specify the timestamp of a query. Queries with `AS OF` do not constrain the timestamp
+selection of other queries. These properties allow clients to circumvent our consistency guarantees, so `AS OF` is left
+out of this document.
+
+## Description
+
+Strict Serializability (in simplified terms) means that you can assign a total order to all transactions in the system,
+and that the total order is constrained by the real time occurrences of those transactions. In other words, if
+transaction A finishes in real time before transaction B starts in real time, then transaction A MUST be ordered before
+transaction B in the total order. Concurrent transaction can be ordered arbitrarily. For a more in depth discussion
+please
+read [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+and chapters 7 and 9 of [Designing Data-Intensive Applications](https://dataintensive.net/).
+
+Materialize can satisfy the constraints of Strict Serializability if all reads and writes occur at specific timestamps,
+and these timestamps are non-decreasing in real-time order, with strict increases for writes following reads. Each
+timestamp is assigned to an event at some real-time moment within the event's bounds (its start and stop). Each
+timestamp transition is made durable before any subsequent response is issued.
+
+We can then use these timestamp to form a total order constrained by real time:
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  write-only, and T2 is write-only.
+
+### Global Timestamp
+
+All reads to any object and all writes to user tables are assigned a timestamp by the Coordinator on a single thread.
+The logic for determining what timestamp to return must satisfy:
+
+- If it's a read, then it is larger than or equal to the `since` of all objects involved in the query.
+- If it's a write, then it is larger than or equal to the `upper` of the table.
+- It is larger than or equal to the global timestamp
+- It is larger than or equal to the timestamp of the most recently completed query.
```

Comment:
maybe i just need to read further along, but why isn the global timestamp the "timestamp of the most recently completed query"?

Reply:
Lines 69 and 70 are redundant, I removed line 69.